src/aap_eda/services/sync_certs.py

```diff
@@ -0,0 +1,214 @@
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+"""Synchronize Certificates with Gateway."""
+import hashlib
+import logging
+from typing import Any, Dict, Optional
+from urllib.parse import urljoin
+
+import requests
+import yaml
+from ansible_base.resource_registry import resource_server
+from django.conf import settings
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+from rest_framework import status
+
+from aap_eda.core import enums, models
+from aap_eda.core.exceptions import GatewayAPIError, MissingCredentials
+
+LOGGER = logging.getLogger(__name__)
+SLUG = "api/gateway/v1/ca_certificates/"
+DEFAULT_TIMEOUT = 30
+SERVICE_TOKEN_HEADER = "X-ANSIBLE-SERVICE-AUTH"
+
+
+class SyncCertificates:
+    """This class synchronizes the certificates with Gateway."""
+
+    def __init__(self, eda_credential_id: int) -> None:
+        self.eda_credential_id: int = eda_credential_id
+        self.gateway_url: str = settings.RESOURCE_SERVER["URL"]
+        self.gateway_ssl_verify: bool = settings.RESOURCE_SERVER.get(
+            "VALIDATE_HTTPS", True
+        )
+
+        self.eda_credential: models.EdaCredential = (
+            models.EdaCredential.objects.get(id=self.eda_credential_id)
+        )
+
+    def update(self) -> None:
+        """Handle creating and updating the certificate in Gateway."""
+        inputs: Dict[str, Any] = yaml.safe_load(
+            self.eda_credential.inputs.get_secret_value()
+        )
+        existing_object: Dict[str, Any] = self._fetch_from_gateway()
+
+        # If the user had a certificate and then they deleted it
+        # remove it from Gateway
+        if existing_object and not inputs["certificate"]:
+            return self.delete(None)
+
+        # If the user has not provided any certificate nothing to do
+        if not inputs["certificate"]:
+            return
+
+        sha256: str = hashlib.sha256(
+            inputs["certificate"].encode("utf-8")
+        ).hexdigest()
+
+        if existing_object.get("sha256", "") != sha256:
+            data: Dict[str, str] = {
+                "name": self.eda_credential.name,
+                "pem_data": inputs["certificate"],
+                "sha256": sha256,
+                "related_id_reference": self._get_remote_id(),
+            }
+            headers: Dict[str, str] = self._prep_headers()
+            if existing_object:
+                slug: str = f"{SLUG}/{existing_object['id']}/"
+                url: str = urljoin(self.gateway_url, slug)
+                response: requests.Response = requests.patch(
```

Comment:
Here we have calls to requests library that can raise exceptions like ConnectionError and are not catch.